crates/phactory/src/prpc_service.rs

```diff
@@ -226,25 +226,33 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> Phactory<Platform>
         let counters = self.runtime_state()?.storage_synchronizer.counters();
         blocks.retain(|b| b.block_header.number >= counters.next_block_number);
 
-        let mut last_block = counters.next_block_number - 1;
+        let last_block = blocks
+            .last()
+            .map(|b| b.block_header.number)
+            .unwrap_or(counters.next_block_number - 1);
+
+        let safe_mode = self.args.safe_mode;
+
         for block in blocks.into_iter() {
             info!("Dispatching block: {}", block.block_header.number);
             let state = self.runtime_state()?;
             state
                 .storage_synchronizer
                 .feed_block(&block, state.chain_storage.inner_mut())
                 .map_err(from_display)?;
+            if safe_mode {
```

Comment:
In this way, we can bypass the logic issue, but if the storage runs out of the memory, it can still fail. A lighter approach is to only sync the block header. When reading the data, we can ask for a storage merkle proof at that block.

Reply:
Good idea. Let me cook a patch.

Reply:
Added two levels of safe mode. Untested yet.

Reply:
Tested in SW mode.